warehouse/email/__init__.py

```diff
@@ -466,6 +466,18 @@ def send_recovery_code_reminder_email(request, user):
     return {"username": user.username}
 
 
+@_email("oidc-provider-added")
+def send_oidc_provider_added_email(request, user, project_name):
+    # We use the request's user, since they're the one triggering the action.
+    return {"username": request.user.username, "project_name": project_name}
+
+
+@_email("oidc-provider-removed")
+def send_oidc_provider_removed_email(request, user, project_name):
+    # We use the request's user, since they're the one triggering the action.
+    return {"username": request.user.username, "project_name": project_name}
```

Comment:
This might be difficult to do in a polymorphic manner, since only `GitHubProvider` has those fields. Something like this should work, though:

* Publisher name: GitHub (via `OIDCProvider.provider_name`
* Publisher specification: foo.yml @ foo/bar (via `str(provider)`)

Reply:
4651ce87